apps/enc.c

```diff
@@ -543,26 +570,92 @@ int enc_main(int argc, char **argv)
         }
     }
 
-    /* Only encrypt/decrypt as we write the file */
-    if (benc != NULL)
-        wbio = BIO_push(benc, wbio);
-
+    /* Read input to buffer BIO, because some ciphers can't be streamed into (usually AEAD) */
+    if ((rbufbio = BIO_new(BIO_s_mem())) == NULL) {
+        BIO_printf(bio_err, "error creating read buffer\n");
+        goto end;
+    }
     for (;;) {
         inl = BIO_read(rbio, (char *)buff, bsize);
         if (inl <= 0)
             break;
-        if (BIO_write(wbio, (char *)buff, inl) != inl) {
+        if (BIO_write(rbufbio, (char *)buff, inl) != inl) {
+            BIO_printf(bio_err, "error writing to read buffer\n");
+            goto end;
+        }
+    }
+    len = BIO_get_mem_data(rbufbio, &p);
+    if (len <= 0) {
+        BIO_printf(bio_err, "error reading input file\n");
+        goto end;
+    }
+
+    /* When decrypting, extract and set final AEAD tag */
+    if (!enc && aeadtaglen > 0) {
+        len -= aeadtaglen;
+        if (len <= 0) {
+            BIO_printf(bio_err, "error reading input file\n");
+            goto end;
+        }
+
+        memcpy(aeadtag, p + len, aeadtaglen);
+
+        if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, aeadtaglen, aeadtag) != 1) {
+            BIO_printf(bio_err, "error setting AEAD tag\n");
+            goto end;
+        }
+    }
+
+    /* Encrypt/decrypt to another buffer BIO, so that output is written only if operation succeeds */
+    if ((wbufbio = BIO_new(BIO_s_mem())) == NULL) {
+        BIO_printf(bio_err, "error creating write buffer\n");
+        goto end;
+    }
+    if (benc != NULL) {
+        wbufbio = BIO_push(benc, wbufbio);
+    }
+    if (BIO_write(wbufbio, (char *)p, len) != len) {
+        BIO_printf(bio_err, "bad decrypt\n");
+        goto end;
+    }
+    if (!BIO_flush(wbufbio)) {
+        BIO_printf(bio_err, "bad decrypt\n");
+        goto end;
+    }
+
+    /* Write output */
+    len = BIO_get_mem_data(wbufbio, &p);
+    if (BIO_write(wbio, p, len) != len) {
+        BIO_printf(bio_err, "error writing output file\n");
+        goto end;
+    }
+
+    /* When encrypting, write AEAD tag */
+    if (enc && aeadtaglen > 0) {
```

Comment:
Just wondering here.  Are there any advantages to writing the tag _before_ the encrypted data rather than after?  Would it simplify the code any?  It would get rid of the `len` checks above.  Since everything is being held in memory, there is complete freedom as to where the various bits go in the output.

Reply:
My experience is protocols tend to put it at the end. It gets computed last and this means in-place encryption and decryption works a bit better. (Things don't shift around.)